### PR TITLE
Implement GPURenderPassEncoder

### PR DESCRIPTION
**Pull request by kunalmohan**
Add webidls for GPURenderPassEncoder and  GPURenderEncoderBase and
implement relevant methods.

r?@kvark

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by kunalmohan** [comment]
```
error[E0432]: unresolved import `crate::dom::bindings::codegen::UnionTypes::GPULoadOpOrRangeEnforcedUnsignedLong`
  --> components/script/dom/gpucommandencoder.rs:13:5
   |
13 |     GPULoadOpOrRangeEnforcedUnsignedLong as GPULongLoad,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GPULoadOpOrRangeEnforcedUnsignedLong` in `dom::bindings::codegen::UnionTypes`
```
This did not show up during my local build :(

The thing is I have this dictionary in `GPUCommandEncoder.webidl`
```
dictionary GPURenderPassDepthStencilAttachmentDescriptor {
    required GPUTextureView attachment;

    required (GPULoadOp or float) depthLoadValue;
    required GPUStoreOp depthStoreOp;
    boolean depthReadOnly = false;

    required (GPULoadOp or GPUStencilValue) stencilLoadValue;
    required GPUStoreOp stencilStoreOp;
    boolean stencilReadOnly = false;
};
```
I need to match on `stencilLoadValue` and `depthLoadValue` for which I need to import their types. Is there someplace else I should import them?

**Comment by jdm** [comment]
This is really strange.
@bors-servo try=linux

**Comment by jdm** [comment]
It makes no sense to me that your local build does not reproduce this build error. My only advice at this point is to do the following:
* make a build error occur in one of gpuadapter.rs by removing an argument from one of the trait methods
* using the path to the generated file that is present in the error message you receive, find the generated UnionTypes.rs that is in the generated bindings' parent directory
* gist that file so we can better understand if GPULoadOpOrRangeEnforcedUnsignedLong is actually present or not

**Comment by kunalmohan** [comment]
@jdm https://gist.github.com/kunalmohan/59b4c4dc2beb3fb789da2ed2507a9bcf

File: components/script/dom/gpurenderpassencoder.rs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    command_encoder: Dom<GPUCommandEncoder>,
+}
+
+impl GPURenderPassEncoder {
+    fn new_inherited(
+        channel: WebGPU,
+        raw_pass: RawPass,
+        parent: &GPUCommandEncoder,
+        device: WebGPUDevice,
+    ) -> GPURenderPassEncoder {
```
Comment (kvark) [inline components/script/dom/gpurenderpassencoder.rs:43]: nit: `-> Self`, can be used in other places too

**Review by kvark — APPROVED** [review]
Looks great! I have a few questions, no blocking concerns.

**Comment by jdm** [comment]
Super weird. I built your branch locally and don't see the build error. What OS are you using?

File: components/script/dom/gpucomputepassencoder.rs [hunk 2]
```diff
@@ -21,28 +22,33 @@ use webgpu::{
...
 #[dom_struct]
 pub struct GPUComputePassEncoder {
     reflector_: Reflector,
     #[ignore_malloc_size_of = "defined in webgpu"]
     channel: WebGPU,
+    device: WebGPUDevice,
     label: DomRefCell<Option<DOMString>>,
     #[ignore_malloc_size_of = "defined in wgpu-core"]
-    raw_pass: RefCell<Option<RawPass>>,
+    raw_pass: DomRefCell<Option<RawPass>>,
```
[the remark this answers is not in the document]
Comment (kunalmohan) [reply components/script/dom/gpucomputepassencoder.rs:36]: It's isn't necessary. RefCell would work too. Its just that DomRefCell has a gives useful output for dynamic borrow failures behind the feature "refcell-backtrace".

File: components/script/dom/gpucomputepassencoder.rs [hunk 3]
```diff
@@ -21,28 +22,33 @@ use webgpu::{
...
-    WebGPU, WebGPURequest,
+    WebGPU, WebGPUDevice, WebGPURequest,
 };
 
 #[dom_struct]
 pub struct GPUComputePassEncoder {
     reflector_: Reflector,
     #[ignore_malloc_size_of = "defined in webgpu"]
     channel: WebGPU,
+    device: WebGPUDevice,
```
[the remark this answers is not in the document]
Comment (kunalmohan) [reply components/script/dom/gpucomputepassencoder.rs:33]: Yep. I'll remove this.

File: components/script/dom/gpurenderpassencoder.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    fn SetPipeline(&self, pipeline: &GPURenderPipeline) {
+        if let Some(raw_pass) = self.raw_pass.borrow_mut().as_mut() {
+            unsafe { wgpu_render::wgpu_render_pass_set_pipeline(raw_pass, pipeline.id().0) };
+        }
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpurenderencoderbase-setindexbuffer
+    fn SetIndexBuffer(&self, buffer: &GPUBuffer, offset: u64, size: u64) {
+        let s;
+        if size == 0 {
```
[the remark this answers is not in the document]
Comment (kunalmohan) [reply components/script/dom/gpurenderpassencoder.rs:185]: I haven't updated wgpu-core recently, so no. I'll do so now.

**Comment by kunalmohan** [comment]
I am on Ubuntu 18.04.

**Comment by jdm** [comment]
As a possible workaround, does anything change if you make a typedef for that union and import the typedef name in rust instead?

**Comment by kunalmohan** [comment]
I have modified that particular type to use typedef, as suggested, to check if the build succeeds.
I haven't yet addressed all the comments from the review yet- wgpu-core is yet to be updated- so please don't merge.

**Comment by kunalmohan** [comment]
So build succeeds now \o/. I don't know why that particular type was throwing the error while the other two didn't (Maybe it was cursed :P).
I'll proceed with this if that's not a problem, and push the other changes too.

**Comment by kunalmohan** [comment]
@kvark PTAL!

**Review by kvark — APPROVED** [review]

**Pull request merged**